cmds/rm/rm.go

```diff
@@ -11,81 +11,166 @@
 //     -i: interactive mode
 //     -v: verbose mode
 //     -R: remove file hierarchies
-//     -r: equivalent to -R
+
 package main
 
 import (
 	"bufio"
+	"errors"
 	"flag"
 	"fmt"
+	"io/ioutil"
 	"os"
-	"path"
+	"path/filepath"
 )
 
-var (
-	flags struct {
-		r bool
-		v bool
-		i bool
+type rmFlags struct {
+	recursive   bool
+	verbose     bool
+	interactive bool
+}
+
+func isEmpty(file string) (bool, error, []os.FileInfo) {
+	fileList, err := ioutil.ReadDir(file)
+	if err != nil {
+		//ignore this boolean return value
+		return false, err, fileList
 	}
-	cmd = "rm [-Rrvi] file..."
-)
+	if len(fileList) == 0 {
+		return true, nil, fileList
+	}
+	return false, nil, fileList
+}
 
-func init() {
-	defUsage := flag.Usage
-	flag.Usage = func() {
-		os.Args[0] = cmd
-		defUsage()
+func interactivePrompt(input *bufio.Scanner, s string, args ...interface{}) (bool, error) {
+	for {
+		fmt.Printf(s+" (y/n)", args...)
+		input.Scan()
+		if input.Err() != nil {
+			return false, input.Err()
+		}
+		switch input.Text() {
+		case "y":
+			return true, nil
+		case "n":
+			return false, nil
+		}
+		fmt.Printf("Please only enter y or n. You entered %s. \n", input.Text())
 	}
-	flag.BoolVar(&flags.i, "i", false, "Interactive mode.")
-	flag.BoolVar(&flags.v, "v", false, "Verbose mode.")
-	flag.BoolVar(&flags.r, "R", false, "Remove file hierarchies")
-	flag.BoolVar(&flags.r, "r", false, "Equivalent to -R.")
-	flag.Parse()
 }
 
-func rm(files []string) error {
-	f := os.Remove
-	if flags.r {
-		f = os.RemoveAll
+func printRemove(flags rmFlags, input *bufio.Scanner, s1 string, s2 string, file string) (bool, error) {
+	if flags.interactive {
+		iVal, err := interactivePrompt(input, s1, file)
+		if err != nil {
+			return false, err
+		}
+		if !iVal {
+			return false, nil
+		}
+	}
+	if err := os.Remove(file); err != nil {
+		return false, err
+	}
+	if flags.verbose {
+		fmt.Printf(s2, file)
 	}
+	return true, nil
+}
 
-	workingPath, err := os.Getwd()
+func rmImplement(file string, flags rmFlags) error {
+	input := bufio.NewScanner(os.Stdin)
+	statval, err := os.Stat(file)
 	if err != nil {
 		return err
 	}
-
-	input := bufio.NewScanner(os.Stdin)
-	for _, file := range files {
-		if flags.i {
-			fmt.Printf("rm: remove '%v'? ", file)
-			input.Scan()
-			if input.Text()[0] != 'y' {
-				continue
+	if statval.IsDir() {
+		if !flags.recursive {
+			return errors.New(fmt.Sprintf("rm : %s : is a directory", file))
+		}
+		if !flags.interactive && !flags.verbose {
+			if err := os.RemoveAll(file); err != nil {
+				return err
 			}
+			return nil
 		}
-
-		if err := f(file); err != nil {
+		empty, err, fileList := isEmpty(file)
```

Comment:
isEmpty isn't adding much value. You can achieve the same in the same number of lines without the indirection:

```go
list, err := ioutil.ReadDir(file)
switch {
case err != nil:
  return err

case len(list) == 0:
  cont, err := ...

case len(list) > 0:
  ...
}
```